# EXERCISE 4.5 

# 01. Import Libraries

In [126]:
# Import pandas and NumPy libraries
import pandas as pd
import numpy as np
import os

# 02. Import Data set from Original Data folder

In [128]:
# Turn the folder path into variable to get access easily. 
path = r'/Users/amritkaur/Desktop/CF Data Immersion/Chapter 4/Achievement 4'

In [129]:
# Import 'products.csv' data set from Orginal Data folder
df_prods = pd.read_csv(os.path.join(path, '02 Data','Original Data', 'products.csv'), index_col = False)

In [130]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [131]:
# Import order_wrangled.csv data set from Prepared Data folder
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [132]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0           0   2539329        1             1                  2   
1           1   2398795        1             2                  3   
2           2    473747        1             3                  3   
3           3   2254736        1             4                  4   
4           4    431534        1             5                  4   

   order_hour_of_day  days_since_last_order  
0                  8                    NaN  
1                  7                   15.0  
2                 12                   21.0  
3                  7                   29.0  
4                 15                   28.0

# 03. Mixed-Type Data

In [134]:
# Create a new Dataframe
df_test = pd.DataFrame()

In [135]:
# Create a mox type column 
df_test['mix'] = ['a', 'b', 1, True]

In [136]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [137]:
# check for mixed type
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


# 04. Missing Data

In [139]:
# Find the missing values in products.csv data set
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [140]:
# To actually view these 16 values, you can create a subset of the dataframe containing only the missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [141]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [142]:
# Check the number of rows in your current dataframe
df_prods.shape

(49693, 5)

In [143]:
# Create a new dataframe of clean data
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [144]:
df_prods_clean

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49677 rows x 5 columns]

In [145]:
df_prods_clean.shape

(49677, 5)

# 05. Duplicate Data

In [147]:
# To check duplicate values
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [148]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [149]:
# create a new dataframe that doesn’t include the duplicates you just identified using the drop_duplicates() function
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [150]:
df_prods_clean_no_dups.shape

(49672, 5)

In [151]:
# export the dataframe into csv file.
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

In [152]:
df_prods_clean_no_dups.shape

(49672, 5)

# TASK 4.5

## Step 2.

In [155]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week  order_hour_of_day  days_since_last_order  
count       3.421083e+06       3.421083e+06           3.214874e+06  
mean        2.776219e+00       1.345202e+01           1.111484e+01  
std         2.046829e+00       4.226088e+00           9.206737e+00  
min         0.000000e+00       0.000000e+00           0.000000e+00  
25%         1.000000e+00       1.000000e+01           4.000000e+00  
50%         3.000000e+00       1.300000e+01           7.000000e+00  
75%         5.000000e+00       1.600000e+01           1.500000e+01  
max         6.000000e+00       2.300000e+01           3.000000e+01

#### Days_Since_last_Order: The "days_since_last_order" column has a lower count compared to other rows, which might indicate that there are missing values in the data. A value of zero in this column indicates that some orders were placed on the same day. The maximum value is 30, which suggests that there might be a limit on the time window within which orders are recorded.
#### Order_Number: The "order_number" column has a maximum value of 100, which could imply that there is a cap on the number of items that can be ordered per transaction.
#### Order_Day_of_the_Week: The "order_day_of_week" column contains values ranging from 0 to 6, where these numbers correspond to the days of the week. This gives a clear indication of the day on which the order was placed.
#### Order_Hour_of_the_Day: The "order_hour_of_day" column has values from 0 to 23, representing the 24 hours in a day. This gives insight into the specific time of day the order was placed (e.g., 0 = midnight, 12 = noon, 23 = 11 PM).

## Step 3

In [158]:
# Check for mixed-type data in your df_ords dataframe.
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [159]:
# There's no mixed-type data in df_ords dataframe.
df_ords.dtypes

Unnamed: 0                 int64
order_id                   int64
user_id                    int64
order_number               int64
order_day_of_week          int64
order_hour_of_day          int64
days_since_last_order    float64
dtype: object

## Step 5

In [161]:
# Check for missing data in dataframe.
df_ords.isnull().sum()

Unnamed: 0                    0
order_id                      0
user_id                       0
order_number                  0
order_day_of_week             0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

### The 'days_since_last_order' column has 206209 missing values. This could happen due to customers' first orders, as these orders wouldn't have any prior order to reference or the missing value for the time gap.

## Step 6

### Address the missing values using an appropriate method.


In [165]:
# To determine the missing values in df_ords dataframe, we need to create a subset dataframe that contains only null values.
df_ords_nan = df_ords[df_ords['days_since_last_order'].isnull() == True]

In [166]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0                 0   2539329        1             1                  2   
11               11   2168274        2             1                  2   
26               26   1374495        3             1                  1   
39               39   3343014        4             1                  6   
45               45   2717275        5             1                  3   
...             ...       ...      ...           ...                ...   
3420930     3420930    969311   206205             1                  4   
3420934     3420934   3189322   206206             1                  3   
3421002     3421002   2166133   206207             1                  6   
3421019     3421019   2227043   206208             1                  1   
3421069     3421069   3154581   206209             1                  3   

         order_hour_of_day  days_since_last_order  
0                        8                    NaN  
11                      11                    NaN  
26                      14                    NaN  
39                      11                    NaN  
45                      12                    NaN  
...                    ...                    ...  
3420930                 12                    NaN  
3420934                 18                    NaN  
3421002                 19                    NaN  
3421019                 15                    NaN  
3421069                 11                    NaN  

[206209 rows x 7 columns]

In [167]:
# To check the 'order_number'column with values of 1 are missing values in the 'days_since_last_order' column, 
# then we need to apply filter the dataframe where 'order number' is '1'.
df_orders_1 = df_ords_nan[df_ords_nan['order_number'] == '1']

In [168]:
# Check if all missing values are in 'days_since_last_order' column.
df_orders_1_missing = df_orders_1['days_since_last_order'].isna().all()

In [169]:
# Print the result of missing values in 'days_since_last_order'.
print(df_orders_1_missing) 

True


In [170]:
# Address the missing values by creating a column that flags order numbers as either True/False depending on if the 'order_number' = 1
df_ords['first_order'] = df_ords['days_since_last_order'].isna()

In [171]:
df_ords

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0                 0   2539329        1             1                  2   
1                 1   2398795        1             2                  3   
2                 2    473747        1             3                  3   
3                 3   2254736        1             4                  4   
4                 4    431534        1             5                  4   
...             ...       ...      ...           ...                ...   
3421078     3421078   2266710   206209            10                  5   
3421079     3421079   1854736   206209            11                  4   
3421080     3421080    626363   206209            12                  1   
3421081     3421081   2977660   206209            13                  1   
3421082     3421082    272231   206209            14                  6   

         order_hour_of_day  days_since_last_order  first_order  
0                        8                    NaN         True  
1                        7                   15.0        False  
2                       12                   21.0        False  
3                        7                   29.0        False  
4                       15                   28.0        False  
...                    ...                    ...          ...  
3421078                 18                   29.0        False  
3421079                 10                   30.0        False  
3421080                 12                   18.0        False  
3421081                 12                    7.0        False  
3421082                 14                   30.0        False  

[3421083 rows x 8 columns]

In [172]:
# Check data-type of new column and update if necessary
df_ords['first_order'].dtype

dtype('bool')

### Now, we have addressed the missing values in days_since_last_order' column appropriately, the next step I would prefer to contact the client to know why these missing values are. However, the data in 'day_since_last_order' column would be helpful for further analysis, so removing or filtering data is not a better choice. 

## Step 7

### Run a check for duplicate values in your df_ords data.

In [176]:
# Check for duplicate values
df_ords_dup = df_ords[df_ords.duplicated()]

In [177]:
df_ords_dup

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, order_day_of_week, order_hour_of_day, days_since_last_order, first_order]
Index: []

### There are no duplicate values in df_ords dataframe. 

## Step 9

### Export the cleaned 'df_ords' data set. 

In [181]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))